### PR TITLE
feat: manual run with json

### PR DESCRIPTION
**Pull request by RonenMars**
## Description
Add an option to send data as JSON to manual run instead of the current key-value.

Add a toggle/dropdown selection between key-value and JSON like this:

image

## Linear Ticket
https://linear.app/autokitteh/issue/UI-778/manual-run-add-json-parameters

## What type of PR is this? (check all applicable)

-   [x] 💡 (feat) - A new feature (non-breaking change which adds functionality)
-   [ ] 🔄 (refactor) - Code Refactoring - A code change that neither fixes a bug nor adds a feature
-   [ ] 🐞 (fix) - Bug Fix (non-breaking change which fixes an issue)
-   [ ] 🏎 (perf) - Optimization
-   [ ] 📄 (docs) - Documentation - Documentation only changes
-   [ ] 📄 (test) - Tests - Adding missing tests or correcting existing tests
-   [ ] 🎨 (style) - Styles - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] ⚙️ (ci) - Continuous Integrations - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] ☑️ (chore) - Chores - Other changes that don't modify src or test files
-   [ ] ↩️ (revert) - Reverts - Reverts a previous commit(s).

### CODE REVIEW

**Comment by RonenMars** [comment]
Add one level JSON validation.

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 1]
```diff
@@ -1,112 +1,243 @@
...
+			if (jsonValue) {
+				try {
+					const parsedJson = JSON.parse(jsonValue);
+					const newParams = Object.entries(parsedJson).map(([key, value]) => ({
+						key,
+						value: JSON.stringify(value),
+					}));
+					setValue("params", newParams, { shouldValidate: true });
+					clearErrors("jsonParams");
+				} catch {
```
Comment (J1za) [inline src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:126]: why do you use try-catch?

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 2]
```diff
@@ -1,112 +1,243 @@
...
+					// If JSON is invalid, keep the current params
+				}
+			}
+		} else {
```
Comment (J1za) [inline src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:130]: without `else`

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 3]
```diff
@@ -1,112 +1,243 @@
...
+				setValue("params", newParams, { shouldValidate: false });
+			} catch {
+				// If JSON is invalid, don't update params
+			}
+		} else {
+			setValue("params", [], { shouldValidate: false });
+		}
+	};
+
+	const toggleEditorMode = () => {
```
Comment (J1za) [inline src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:112]: I don't like that it changes from JSON to inputs , I see an empty `""`
image
image

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 4]
```diff
@@ -1,112 +1,243 @@
...
+
+	const convertParamsToJson = (currentParams: { key: string; value: string }[]) => {
+		const jsonObject = Object.fromEntries(
+			currentParams.map((param: { key: string; value: any }) => {
+				try {
+					return [param.key, JSON.parse(param.value)];
+				} catch {
+					return [param.key, param.value];
+				}
+			})
```
Comment (J1za) [inline src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:41]:
```suggestion
			currentParams.map(({ key, value }) => {
				try {
					return [key, JSON.parse(value)];
				} catch {
					return [key, value];
				}
			})
```

File: src/locales/en/deployments/translation.json [hunk 5]
```diff
@@ -53,7 +53,8 @@
 			"executionFailedExtended": "Session execution failed, projectID - {{projectId}}, error - {{error}}",
 			"executionSucceed": "Session execution succeed for session ID",
 			"missingLastDeployment": "Missing last deployment",
-			"noFunctionsFound": "No functions found"
+			"noFunctionsFound": "No functions found",
```
Comment (J1za) [inline src/locales/en/deployments/translation.json:56]: Where is this used?

File: src/validations/manualRun.schema.ts [hunk 6]
```diff
@@ -7,55 +7,76 @@ const selectSchema = z.object({
 });
 
 const paramSchema = z.object({
-	key: z.string(),
-	value: z.string(),
+	key: z.string().min(1, i18n.t("keyIsRequired", { ns: "validations" })),
```
Comment (J1za) [inline src/validations/manualRun.schema.ts:10]: i18n working after ` i18n.on("initialized", () =>`

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 7]
```diff
@@ -1,112 +1,243 @@
...
-	const errors = formState.errors?.params as FieldErrors<FieldValues> & ManualFormParamsErrors;
+	useEffect(() => {
+		if (isJson) {
+			const currentParams = getValues("params");
+			convertParamsToJson(currentParams);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [projectId]);
+
+	const [useJsonEditor, setUseJsonEditor] = useState(isJson);
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:54]: done

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 8]
```diff
@@ -1,112 +1,243 @@
...
+
+	const errors = formState.errors as FieldErrors<FieldValues> & ManualFormParamsErrors;
+
+	const validateParams = () => {
+		const params = getValues("params");
+		let isValid = true;
+
+		params.forEach((param: { key: string; value: string }, index: number) => {
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:62]: fixed

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 9]
```diff
@@ -1,112 +1,243 @@
...
+			if (!param.key.trim()) {
+				setError(`params.${index}.key`, {
+					type: "manual",
+					message: t("keyIsRequired"),
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:66]: Fixed

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 10]
```diff
@@ -1,112 +1,243 @@
...
+				const newParams = Object.entries(parsedJson).map(([key, value]) => {
+					if (typeof value === "object") return { key, value: JSON.stringify(value) };
+
+					return { key, value: String(value) };
+				});
+				setValue("params", newParams, { shouldValidate: false });
+			} catch {
+				// If JSON is invalid, don't update params
+			}
+		} else {
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:107]: refactored without using the else

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 11]
```diff
@@ -1,112 +1,243 @@
...
+		if (value) {
+			try {
+				const parsedJson = JSON.parse(value);
+				const newParams = Object.entries(parsedJson).map(([key, value]) => {
+					if (typeof value === "object") return { key, value: JSON.stringify(value) };
+
+					return { key, value: String(value) };
+				});
+				setValue("params", newParams, { shouldValidate: false });
+			} catch {
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:104]: Refactored to use try catch only in the parse function

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx (hunk 1 above)
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:126]: refactored

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx (hunk 2 above)
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:130]: refactored

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx (hunk 3 above)
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:112]: fixed

File: src/locales/en/deployments/translation.json (hunk 5 above)
Comment (RonenMars) [reply src/locales/en/deployments/translation.json:56]: removed

File: src/validations/manualRun.schema.ts (hunk 6 above)
Comment (RonenMars) [reply src/validations/manualRun.schema.ts]: fixed

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx (hunk 4 above)
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:44]: refactored to work without try-catch

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 12]
```diff
@@ -1,112 +1,265 @@
...
+			return JSON.parse(value);
+		} catch {
+			return null;
+		}
 	};
 
-	const errors = formState.errors?.params as FieldErrors<FieldValues> & ManualFormParamsErrors;
+	const convertParamsToJson = (currentParams: ManualRunParam[]) => {
+		const jsonObject = Object.fromEntries(
+			currentParams.map(({ key, value }) => [key, safeJsonParse(value) ?? value])
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:46]: [commented](https://github.com/autokitteh/web-platform/pull/522#discussion_r1809962020)

File: src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx [hunk 13]
```diff
@@ -1,112 +1,265 @@
...
+		setValue("params", newParams, { shouldValidate: true });
+		clearErrors("jsonParams");
+	};
 
-				<div className="cursor-pointer" title={t("titleParams")}>
-					<InfoIcon className="fill-white" />
+	const handleParamsToJsonConversion = () => {
+		const currentParams: ManualRunParam[] = getValues("params");
+		const jsonObject = currentParams.reduce<ManualRunJsonObject>((acc, { key, value }) => {
+			acc[key] = safeJsonParse(value) ?? value;
```
[the remark this answers is not in the document]
Comment (RonenMars) [reply src/components/organisms/topbar/project/manualRun/manualRunParamsForm.tsx:116]: [commented](https://github.com/autokitteh/web-platform/pull/522#discussion_r1809962020)

**Pull request merged**